src/coreclr/jit/importer.cpp

```diff
@@ -1311,7 +1311,42 @@ GenTree* Compiler::impAssignStructPtr(GenTree*             destAddr,
             {
                 if (callConvIsInstanceMethodCallConv(srcCall->GetUnmanagedCallConv()))
                 {
+#ifdef TARGET_X86
+                    if (srcCall->GetUnmanagedCallConv() != CorInfoCallConvExtension::Thiscall)
+                    {
+                        // The argument list has already been reversed.
+                        // Insert the return buffer as the second-to-last node
+                        // so it will be pushed on to the stack after the user args but before the native this arg
+                        // as required by the native ABI.
+                        GenTreeCall::Use* lastArg = srcCall->gtCallArgs;
+                        if (lastArg == nullptr)
+                        {
+                            srcCall->gtCallArgs = gtPrependNewCallArg(destAddr, srcCall->gtCallArgs);
+                        }
+                        else
+                        {
+                            if (lastArg->GetNext() != nullptr)
+                            {
+                                for (; lastArg->GetNext()->GetNext() != nullptr; lastArg = lastArg->GetNext())
```

Comment:
> for (; lastArg->GetNext()->GetNext() != nullptr; lastArg = lastArg->GetNext()) [](start = 31, length = 79)

I don't know the JIT data structures, but this seems weird to me. Is the `lastArg` guaranteed to never be null now? Additionally is `lastArg->GetNext()` as the invariant never going to be null?

Reply:
I think I need to add an assert for `lastArg` not being null in this case. Since the function pointer is an instance method, there's a requirement that there's at least one (pointer-sized) parameter. This is asserted for `thiscall`, but not for all instance methods. I'll go find the place where it's asserted for `thiscall` and update it to also check other "member function" calling conventions.

Reply:
The if right above this case checks if `lastArg` is null, so it won't be null in this `else` condition.

Reply:
@jkoritzinsky But it is a `for` loop with a `lastArg = lastArg->GetNext()`? Which means upon iteration the next `lastArg` isn't what was checked in the above `if`. As I mentioned perhaps I am unclear on JIT data structure invariants, but it seems that a `GetNext()` doesn't make a lot of guarantees about the complete state of its returned value.

Reply:
Good point. I missed that. I'll take another look at this.

Reply:
Rewritten in the most recent commit to correctly do the traversal.